### DISCUSSION TITLE
Directions Q&A - What's New: A Modernized Approach to Email Communications In Business Central

### DISCUSSION THREAD

**Discussion by darjoo** [microsoft/ALAppExtensions#9768] [Q&A] [answered]
Hi all, 

You can find the questions that were asked at the Directions Q&A and the answers here. If you have further questions, feel free to ask them :)

-------------------------

Is this available with Business Central on-premises? Also, does it work on hosted or on-premises Exchange, or only with Office 365? 
> There is full feature parity between Business Central online and on-premises. From version 17.1, both versions offer the same email capabilities. You will be able to connect to hosted Exchange and Exchange on-premises, though you may have to use different connectors. 

How do you handle attachments and database size? 

>The Email module keeps all attachments until the email is deleted. You can use the Retention Policy module to specify how often you want to clean up old messages. 

Is it possible to have a scenario with a "default" email address, but change the sender before sending the mail? 

>Not a scenario per se, but the user can change the sender if the mail is displayed in the editor. 

Is it possible (without coding) to suppress the HTML editor in some scenarios and send the mail directly? 

>Whether the email editor is displayed is based on the Document Sending Profile configuration. Currently, this cannot be controlled on a scenario level. 

Is this available from version 17.0? I understood something about version 19.0? On prem and cloud a difference? 

>The new email capabilities are available in both Business Central online and on-premises from version 17.1. We will retire the legacy SMTP capabilities in version 19.0. 

Are there any special permissions for using email? 

>The new system permission set named EMAIL SETUP can be assigned explicitly to users or user groups to allow them to configure email accounts and scenarios.  

Is it possible to set up a job queue that will move all new, for example, sales orders to the Email Outbox, and then send all invoices at the same time? 

>We don’t currently offer the ability to delay sending. This would be a great addition to the new email story. 

Is it possible to use email templates if you send according a scenario? 

>Unfortunately, we currently don’t support email templates per scenario. 

Will the new mail functionality work with batch sending of mails? / 
Can you send emails in batches? For example, to send all the sales invoices in one go? 

>We don’t currently support batch sending, however, that could easily be added because the module is ready. We’re hoping to see a code contribution to enable such capabilities. 

Could you use the rich text editor for table/page fields that are not related to the Email module? 

>For now, the rich text editor is exclusively available to the email story, but we’re working on making it available as a general-purpose control in Business Central. 

Do you know if we will be able to list all emails related to a customer at some point, for example, from customer or contact cards? 

>Currently, we have no plans to support that. However, we would be happy to process a pull request that enables that scenario. You could also register this as an idea on aka.ms/bcideas. Alternatively, this can already be done by integrating with Microsoft Sales. 

Is the email after the send in the outlook-history? 

>Yes, it will also be in the Sent Items folder in Outlook. 

Is it possible to have extra attachments, such as order confirmations? 

>Yes, you can attach such documents, although the user must do so either manually or via an extension.  

Can you insert a footer with usernames and company logos? 

>There is no support for that now, but we hope that such capabilities will get added either by us or our community. 

Does this functionality alter email logging? 

>No. 

Can you mix and match connectors, for example, if you want all invoices to go from a generic account and service orders from the user account? 

>Yes, you can mix and match connectors. Email scenarios do not depend on the connector used for the account. 

If you attach other files to a sales order, would it also be added in the email attachments? 

>No, by default the attachment is the sales order in PDF format. 

Where is the To address received from? 

>There have been no changes to that logic. The To address comes from the document being processed. 

Is the permissions concept 0 or 1, for privacy reasons, meaning not everyone can read every email? 

>Currently, only the person who sent the email can read it on the Sent Items page. System administrators can view a list of all emails that were sent, which is useful for troubleshooting, but they cannot open the emails. 

Can you add a standard PDF to a sales order, for example, a terms and conditions document? 

>Yes, you can attach such documents, although the user must do so either manually or via an extension. 

Is the email body rich text or HTML? And, if the latter, is HTML5 supported? 

>All details about the editor can be found here: https://github.com/microsoft/roosterjs

**Comment by fvet** [comment]
> How do you handle attachments and database size?
> 
> The Email module keeps all attachments until the email is deleted. You can use the Retention Policy module to specify how often you want to clean up old messages.
> 

Our customers want to keep track of the email history and not apply a retention policy, since they would loose email communicatoin history. If the sent email attachments are stored as blob, will they count against the database size. Why did'nt you implement 'Media*' support for this module? Would that not count towards the database size?

-------

Can we post new ideas here on how we see the email module evolving? There's a lot of content of Yammer (https://www.yammer.com/dynamicsnavdev/#/Threads/show?threadId=908516312195072), a lot of related ideas on Experience (https://experience.dynamics.com/ideas/idea/?ideaid=dbfcab70-e531-eb11-8441-0003ff68e4a1), so not sure when to use Github discussions...

In short, the next step we'd like to see is a 'timeline' feature on the entities factboxarea (master data / sales documents / ...) with a track history of all chats (comments / notes), documents (attachments / record links / incoming documents custom tables) and **communication** (sent emails) (Similar to the Unified Interface of other Dynamics products)

**Reply by JesperSchulz** [reply]
Sorry, I wasn't informed about this post. I think I need to adjust my notification settings! I've provided a short plan below and we're of course all ears wrt. new, even better ideas. Just keep posting them, and we'll try to pick them up :-)

**Comment by JesperSchulz** [comment, marked as answer]
Short update on the work we've got in progress with regards to the new email module, which I expect we'll release piece by piece throughout this year. Of course we're listening to all your feedback and wishes, as posted on Yammer, BCIdeas, etc. 

For version 18.0, we expect to enable you to establish a link between an email and source entities (documents, customers, contacts, etc.).
After that release, we're going to uptake this capability in the BaseApp, by enabling to send mails from these entities with a link established. Also, we're going to make it possible to see those mails again from the source entity (e.g. show related mails from within a posted sales order).
Next, we going to enable to forward documents which are attached to the entities in mails (e.g. send mail from posted sales order, which has attachments from the posted sales order).
And of course we need to fix, that we used blob instead of media - that's simply a bug from our side.

As email isn't my team's main investment for this particular release, we're going to try to fit these things in "on the side", but it is my hope that all of the above will come drizzling after the next major and up until the 19 release.

I hope that gives some insights into our plans for the near future :-)

**Reply by fvet** [reply]
Thanks for the update! Looking forward to the next batch of email improvements ;)

**Reply by pri-kise** [reply]
Thanks for the summary. I'm really looking forward to these improvements.
@JesperSchulz could you ask for adding the Recordreference(s) to the events like OnBeforeEmailFileInternal (in CU codeunit 260 "Document-Mailing")  or will there be an function inside the TempEmailItem to retrieve the SourceRecord(s)?

**Reply by JesperSchulz** [reply]
We're going to have a function on the email module, which looks like this:

```
    ///<summary>
    /// Gets the sent emails related to a record.
    ///</summary>
    ///<param name="TableId">The table ID of the record.</param>
    ///<param name="SystemId">The system ID of the record.</param>
    ///<returns>The sent email related to a record.<returns>
    procedure GetSentEmailsForRecord(TableId: Integer; SystemId: Guid) ResultSentEmails: Record "Sent Email" temporary;
```

It's the other way around, but I assume it will do? Would be easy to add a function for the other direction as well (get all source records for an email).

**Reply by pri-kise** [reply]
Maybe my english wasn't good enough.
I was taking about the procedures in the BaseApp before the email is sent via the new email module.
If i look into the code of the current BC18 preview, i was surprised that at the moment only one reference is saved to the TempEmailItem:
```al
procedure EmailFile(AttachmentStream: Instream; AttachmentName: Text; HtmlBodyFilePath: Text[250]; PostedDocNo: Code[20]; ToEmailAddress: Text[250]; EmailDocName: Text[250]; HideDialog: Boolean; ReportUsage: Integer; SourceReference: RecordRef): Boolean
var
    TempEmailItem: Record "Email Item" temporary;
begin
    TempEmailItem."Source Table" := SourceReference.Number();
    TempEmailItem."Source System Id" := SourceReference.Field(SourceReference.SystemIdNo()).Value();
....
```

Shouldn't that procedure handle a RecordReference containing multiple Records when this function is called from `TrySendToEMailGroupedMultipleSelection` (in the table "Document Sending Profile").

I don't know what's the best way to capture multiple references, but maybe you can add another global variable, similar to the pattern that is used for the new Attachments?
And then add a procedure to the e-mail Item to retrieve the RecordRef...

@JesperSchulz are you still not receiving e-mail notifications for this discussion or haven't you got the time to check my reply?

(Related to Issue #9005)

**Reply by JesperSchulz** [reply]
I was indeed not receiving any notifications, but luckily your issue was mentioned more places than just here. I would think your request has been addressed in the 18.1 release?

![image](https://user-images.githubusercontent.com/19796701/116979439-4ae52c80-acc5-11eb-97bf-dcd33ec77348.png)

If not, please let me know. Try replying to this comment for me to verify, if I get notified now. If not, please use issues instead :-)

**Reply by pri-kise** [reply]
Yes, i checked this and it looks fine.
If you have seen this issue earlier obsoleting of those fields could have been avoided.
I don't know whats the better channel to contact about application errors in preview versions.

**Reply by JesperSchulz** [reply]
Notifications work now - all good!
Sure, would have been good to avoid this extra roundtrip, but in this case, obsoletion wasn't that bad, as the fields were internal - meaning we can remove them at will. This will get cleaned up quickly.
About providing feedback on preview versions, this forum (issues or Q&A) or Yammer are preferred. I'm looking at how we can involve our partner channel earlier in development, but it's a long road. For now, we'll have to manage with what we've got.

**Comment by stgerber** [comment]
Hi,
We want to send a meeting request from AL. How can we do that ?
Many thanks for your help
